### DISCUSSION TITLE
OTLP | Compile OTLP lib

### DISCUSSION THREAD

**Discussion by yonatan555** [open-telemetry/opentelemetry-cpp#1763] [Q&A]
Hi, i installed otlp through vcpkg on linux server and used this command to install otlp
./vcpkg install opentelemetry -cpp[otlp,core, Prometheus].

And I tried to work and compile one of the examples on the  otlp lib, that located under examples/otlp/grpc_metrics_main
And I got an undefined reference- i don't what I have done wrong.

This my makefile I used.

`CC = g++
TARGET = main

CXXFILES = main.cpp
CXXFLAGS = -O3 -o $(TARGET) -D_GNU_SOURCE -L./ -L./metrics_foo_library -I./ -I./opentelemetry -I./metrics_foo_library
LIBS = -lopentelemetry_common -lopentelemetry_exporter_in_memory -lopentelemetry_exporter_ostream_logs -lopentelemetry_exporter_ostream_span -lopentelemetry_otlp_recordable -lopentelemetry_exporter_ostream_metrics_deprecated -lopentelemetry_http_client_curl -lopentelemetry_exporter_otlp_grpc -lopentelemetry_exporter_otlp_grpc_client -lopentelemetry_exporter_otlp_grpc_log  -lopentelemetry_exporter_otlp_http_client -lopentelemetry_exporter_otlp_http -lopentelemetry_exporter_otlp_http_log -lopentelemetry_logs -lopentelemetry_metrics_deprecated -lopentelemetry_proto -lopentelemetry_resources -lopentelemetry_trace -lopentelemetry_version -lprometheus_exporter_deprecated  -lpthread  -lfoo_library -ldl

all:
        $(CC) $(CXXFILES) $(LIBS) $(CXXFLAGS)

clean:
        $(RM) $(TARGET)
`

`/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: /tmp/cc5cO804.o: in function `main':
main.cpp:(.text.startup+0x96): undefined reference to `opentelemetry::v1::exporter::otlp::OtlpGrpcMetricExporterFactory::Create(opentelemetry::v1::exporter::otlp::OtlpGrpcMetricExporterOptions const&)'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: main.cpp:(.text.startup+0x109): undefined reference to `opentelemetry::v1::sdk::metrics::PeriodicExportingMetricReader::PeriodicExportingMetricReader(std::unique_ptr<opentelemetry::v1::sdk::metrics::PushMetricExporter, std::default_delete<opentelemetry::v1::sdk::metrics::PushMetricExporter> >, opentelemetry::v1::sdk::metrics::PeriodicExportingMetricReaderOptions const&)'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: main.cpp:(.text.startup+0x1c6): undefined reference to `opentelemetry::v1::sdk::metrics::MeterProvider::MeterProvider(std::unique_ptr<opentelemetry::v1::sdk::metrics::ViewRegistry, std::default_delete<opentelemetry::v1::sdk::metrics::ViewRegistry> >, opentelemetry::v1::sdk::resource::Resource)'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: main.cpp:(.text.startup+0x2ff): undefined reference to `opentelemetry::v1::sdk::metrics::MeterProvider::AddMetricReader(std::shared_ptr<opentelemetry::v1::sdk::metrics::MetricReader>)'
collect2: error: ld returned 1 exit status
make: *** [makefile:9: all] Error 1
`

**Comment by lalitb** [comment]
vcpkg port for opentelemetry-cpp is not officially maintained/delivered through this repo (as mentioned here - https://github.com/open-telemetry/opentelemetry-cpp/blob/main/INSTALL.md#using-package-managers) , moving this to discussions.

**Comment by lalitb** [comment]
Looking into the compile flags in make file, you seems to be linking with `opentelemetry_metrics_deprecated` and `prometheus_exporter_deprecated` which is the older version of metrics implementation. Try with `opentelemetry_metrics` and  `prometheus_exporter`  instead.

**Reply by yonatan555** [reply]
where can I configure to use the opentelemetry_metrics ,prometheus_exporter  instead in the make file?

**Reply by lalitb** [reply]
Here:
``
CXXFLAGS = -O3 -o $(TARGET) -D_GNU_SOURCE -L./ -L./metrics_foo_library -I./ -I./opentelemetry -I./metrics_foo_library
LIBS = -lopentelemetry_common -lopentelemetry_exporter_in_memory -lopentelemetry_exporter_ostream_logs -lopentelemetry_exporter_ostream_span -lopentelemetry_otlp_recordable -lopentelemetry_exporter_ostream_metrics_deprecated -lopentelemetry_http_client_curl -lopentelemetry_exporter_otlp_grpc -lopentelemetry_exporter_otlp_grpc_client -lopentelemetry_exporter_otlp_grpc_log -lopentelemetry_exporter_otlp_http_client -lopentelemetry_exporter_otlp_http -lopentelemetry_exporter_otlp_http_log -lopentelemetry_logs -lopentelemetry_metrics_deprecated -lopentelemetry_proto -lopentelemetry_resources -lopentelemetry_trace -lopentelemetry_version -lprometheus_exporter_deprecated -lpthread -lfoo_library -ldl
``

**Reply by lalitb** [reply]
Ok, I think the problem is with the vcpkg for otel-cpp, it is still generating old metrics targets. This needs to be fixed in vcpkg.

**Reply by yonatan555** [reply]
Thanks for your comment, so if I want to use in the new version I need to build the project with one of the options,like with cmke, that should build otlp proj, with the new version.

**Reply by mks26** [reply]
On Linux, if you want to build opentelemetry-cpp sdk with OTLP/GRPC exporter,  first you build gRPC C++ :
Quick start guide to build grpc :  https://grpc.io/docs/languages/cpp/quickstart/

After gRPC C++ is build then set the -DCMAKE_PREFIX_PATH to gRPC install directory while building  opentelemetry-cpp WITH_OTLP.

**Reply by lalitb** [reply]
>  so if I want to use in the new version I need to build the project with one of the options,like with cmke, that should build otlp proj, with the new version.

Yes, that's correct.

**Reply by yonatan555** [reply]
> On Linux, if you want to build opentelemetry-cpp sdk with OTLP/GRPC exporter, first you build gRPC C++ : Quick start guide to build grpc : https://grpc.io/docs/languages/cpp/quickstart/
> 
> After gRPC C++ is build then set the -DCMAKE_PREFIX_PATH to gRPC install directory while building opentelemetry-cpp WITH_OTLP.
 
Ok thank you for the answer.
About the 2nd part,  so if grpc dir is located under /home/admin/grpc/

So The cmake command in opentelemetry-cpp/bud  dir should be:
cmake -DWITH_OTLP=ON -DWITH_OLTP_GRPC=ON -DWITH_PROMETHEUS=ON -DCMAKE_PERFIX_PATH=/home/admin/grpc/  ..

Thanks.

**Reply by mks26** [reply]
Yes, that's correct.

**Reply by yonatan555** [reply]
ok thanks, I have a lot of problems with the build:
I didnt add the whole output, but this is a part.

`/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: /root/vcpkg/packages/curl_x64-linux/lib/libcurl.a(content_encoding.c.o): in function `deflate_unencode_write':
content_encoding.c:(.text+0x8c0): undefined reference to `inflateEnd'

/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x64b1): undefined reference to `X509_free'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x64c4): undefined reference to `OCSP_response_status_str'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x6522): undefined reference to `BIO_free'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x6557): undefined reference to `X509_verify_cert_error_string'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x6647): undefined reference to `OCSP_BASICRESP_free'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x664f): undefined reference to `OCSP_RESPONSE_free'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x6686): undefined reference to `ERR_get_error'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x66a9): undefined reference to `ERR_error_string_n'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x6704): undefined reference to `X509_free'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x6776): undefined reference to `ERR_get_error'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: openssl.c:(.text+0x6799): undefined reference to `ERR_error_string_n'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: /root/vcpkg/packages/curl_x64-linux/lib/libcurl.a(openssl.c.o): in function `ossl_session_free':
openssl.c:(.text+0x4a1): undefined reference to `SSL_SESSION_free'

`

**Reply by owent** [reply]
You need link openssl before libcurl.

**Reply by yonatan555** [reply]
how it is can be done?

**Reply by mks26** [reply]
Did you try to build gRPC? if yes in that case you don't need to use "vcpkg" you just point gRPC install directory while building opentelemetry-cpp with OTLP.

**Reply by yonatan555** [reply]
yes I build gRPC, so remove the vcpkg 
and how to point to grpc install dir using this flag:
-DCMAKE_PERFIX_PATH=/root/grpc/ 
?

Thanks

**Reply by mks26** [reply]
Yes, that's right.

**Reply by owent** [reply]
> how it is can be done?

That depends how you compiling libcurl and gRPC, the most case you can link openssl by `-lssl -lcrypto`.

**Reply by yonatan555** [reply]
I compiiled grpc and follow this guide https://grpc.io/docs/languages/cpp/quickstart/ 
I bringed the all lib*.a after compilation, to opentelemetry project and got alot of errors. Idk what I need to do more. 

that is my make file to run the main - the main is an example from otlp_grpc_metrics repository
please I need your herp.

`
CC = g++
TARGET = main

CXXFILES = main.cpp
CXXFLAGS = -O3 -o $(TARGET) -D_GNU_SOURCE -L./ -L./lib -L./metrics_foo_library -I./ -I./opentelemetry -I./metrics_foo_library -I./lib
LIBS = -lopentelemetry_common -lopentelemetry_exporter_in_memory -lopentelemetry_exporter_ostream_metrics -lopentelemetry_exporter_ostream_span -lopentelemetry_exporter_otlp_grpc -lopentelemetry_exporter_otlp_grpc_client -lopentelemetry_exporter_otlp_grpc_log -lopentelemetry_exporter_otlp_grpc_metrics -lopentelemetry_exporter_otlp_http -lopentelemetry_exporter_otlp_http_client -lopentelemetry_exporter_otlp_grpc_client -lopentelemetry_exporter_otlp_http_metric -lopentelemetry_exporter_otlp_grpc_metrics -lopentelemetry_exporter_otlp_http -lopentelemetry_exporter_otlp_http_client -lopentelemetry_exporter_otlp_http_metric -lopentelemetry_exporter_prometheus -lopentelemetry_http_client_curl -lopentelemetry_metrics -lopentelemetry_otlp_recordable  -lopentelemetry_proto -lopentelemetry_resources -lopentelemetry_trace -lopentelemetry_version -lprometheus-cpp-core -lprometheus-cpp-pull -lprometheus-cpp-push -laddress_sorting -lgpr -lgrpc -lgrpc++ -lgrpc++_alts -lgrpc++_error_details -lgrpc_plugin_support -lgrpcpp_channelz -lgrpc++_reflection -lgrpc_unsecure  -lupb -labsl_bad_any_cast_impl -labsl_bad_optional_access -labsl_bad_variant_access -labsl_base -labsl_city -labsl_civil_time -labsl_cord -labsl_cord_internal -labsl_cordz_functions -labsl_cordz_handle -labsl_cordz_info -labsl_cordz_sample_token -labsl_debugging_internal -labsl_demangle_internal -labsl_examine_stack -labsl_exponential_biased -labsl_failure_signal_handler -labsl_flags -labsl_flags_commandlineflag -labsl_flags_commandlineflag_internal -labsl_flags_config -labsl_flags_internal -labsl_flags_marshalling -labsl_flags_parse -labsl_flags_private_handle_accessor -labsl_flags_program_name -labsl_flags_reflection -labsl_flags_usage -labsl_flags_usage_internal -labsl_graphcycles_internal -labsl_hash -labsl_hashtablez_sampler -labsl_int128 -labsl_leak_check -labsl_log_severity -labsl_low_level_hash -labsl_malloc_internal -labsl_periodic_sampler -labsl_random_distributions -labsl_random_internal_distribution_test_util -labsl_random_internal_platform -labsl_random_internal_pool_urbg -labsl_random_internal_randen -labsl_random_internal_randen_hwaes -labsl_random_internal_randen_hwaes_impl -labsl_random_internal_randen_slow -labsl_random_internal_seed_material -labsl_random_seed_gen_exception -labsl_random_seed_sequences -labsl_raw_hash_set -labsl_raw_logging_internal -labsl_scoped_set_env -labsl_spinlock_wait -labsl_stacktrace -labsl_status -labsl_statusor -labsl_strerror -labsl_str_format_internal -labsl_strings -labsl_strings_internal -labsl_symbolize -labsl_synchronization -labsl_throw_delegate -labsl_time -labsl_time_zone -laddress_sorting -lcares -lcrypto -lpthread  -lfoo_library -ldl
all:
        $(CC) $(CXXFILES) $(LIBS) $(CXXFLAGS)

clean:
        $(RM) $(TARGET)
`

**Reply by owent** [reply]
My suggestion is build examples of otel-cpp with cmake, and just copy the flags and link names from generated `flags.cmake` and `links.cmake` of examples.

**Reply by yonatan555** [reply]
How exactly I can do it?
 How to link these (I am new in cmake, and all this)

**Reply by owent** [reply]
Oh, I just find you install otel-cpp by vcpkg.Here is a example to use it in cmake build system.

## test.cpp

```cpp
int main () {
  return 0;
}
```

## CMakeLists.txt

```cmake
cmake_minimum_required(VERSION 3.16.0)
project(test-otel-cpp)

find_package(Threads)
find_package(OpenSSL)
find_package(CURL)
find_package(protobuf)
find_package(absl)
find_package(grpc)
find_package(nlohmann_json)
find_package(prometheus-cpp)
find_package(opentelemetry)

add_executable(test-exec  test.cpp)
target_link_libraries(test-exec
    // Maybe any other components you need
    opentelemetry-cpp::otlp_http_exporter
    opentelemetry-cpp::otlp_grpc_exporter
    opentelemetry-cpp::prometheus_exporter
    opentelemetry-cpp::otlp_http_metric_exporter
    opentelemetry-cpp::otlp_grpc_metrics_exporter
    opentelemetry-cpp::otlp_http_log_exporter
    opentelemetry-cpp::otlp_grpc_log_exporter
    opentelemetry-cpp::ostream_log_exporter
    opentelemetry-cpp::ostream_span_exporter
    opentelemetry-cpp::ostream_metrics_exporter
    opentelemetry-cpp::resources
    opentelemetry-cpp::metrics
    opentelemetry-cpp::sdk
    opentelemetry-cpp::api
    Threads::Threads
)
```

## Configure it with vcpkg

```bash
# Unix Makefiles is only available on Unix like systems, or using Ninja if you have install it.
mkdir <Build directory> && cd <Build directory>
cmake <The directory contains CMakeLists.txt> -G "Unix Makefiles"  -DCMAKE_TOOLCHAIN_FILE=<The vcpkg installation directory>/scripts/buildsystems/vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x64-linux

# Build
cmake --build <Build directory> -j # or cd <Build directory> && make -j

find <Build directory> -name "flags.make"
find <Build directory> -name "link.txt"

```

Then copy the flags and links above into your `Makefile`.

**Comment by yonatan555** [comment]
can be closed, succeussed
